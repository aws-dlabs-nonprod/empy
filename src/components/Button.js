import React, { Component } from 'react';
import styled from 'styled-components';
import Media from '@style/media';
import svgSprite from '@style/svg';

import { StyledLoader } from '@components/Loader';
import Tip, { StyledTip } from '@components/ButtonTip';

class Button extends Component {

    static defaultProps = {
        tip: null,
        toggleTip: null,
        tipPosition: 'left',
        isToggled: false,
        disabled: false
    };

    constructor(props) {
        super(props);
    }

    handleClick = (e) => {
        const { onClick } = this.props;

        if (onClick) {
            e.preventDefault();
            this.props.onClick();
        }
    }

    render() {
        const {
            children,
            className,
            icon,
            toggleIcon,
            tip,
            toggleTip,
            isToggled,
            disabled,
            tipPosition,
            tipBelow,
            theme,
            ...rest
        } = this.props;

        const tipContent = isToggled ? (toggleTip || tip) : tip;
        const iconOption = isToggled ? (toggleIcon || icon) : icon;

        const Component = rest.href ? 'a' : 'button';

        return (
            <Component { ...rest } disabled={ disabled } aria-label={ tipContent } className={ className } onClick={ this.handleClick }>
                { iconOption &&
                    <div dangerouslySetInnerHTML={{ __html: svgSprite(iconOption) }}></div>
                }

                { children }

                { tipContent &&
                <Tip theme={ theme } tipPosition={ tipPosition } tipBelow={ tipBelow }>
                    <p>{ tipContent }</p>
                </Tip>
                }
            </Component>
        );
    }
}

const BaseButton = styled(Button)`
    background: ${props => props.theme.colourPrimary};
    border-radius: ${props => props.theme.radius};
    box-sizing: border-box;
    color: ${props => props.theme.colourForeground};
    cursor: pointer;
    display: flex;
    justify-content: center;
    align-items: center;
    font-size: 1.6rem;
    height: 4.8rem;
    border: ${props => `0.1rem solid ${props.theme.colourPrimary}`};
    margin: 0 0 2rem 0;
    outline: none;
    padding: 0 1.8rem;
    position: relative;
    width: 100%;
    z-index: 3;
    transition: background 0.2s ease, color 0.2s ease;

    ${Media.desktop`
        font-size: 1.8rem;
        padding: 0 2rem;
        width: 100%;

        &:hover {
            background: ${props => props.theme.colourPrimaryHover};
            
            svg {
                path {
                    fill: ${props => props.theme.colourForeground};
                }
            }

            ${StyledTip} {
                opacity: ${props => props.tip ? 1 : 0};
                transform: translateY(0);
            }
        }
    `}
`;

const SecondaryButton = styled(BaseButton)`
    background: ${props => props.theme.colourForeground};
    color: ${props => props.theme.textPrimary};

    ${Media.desktop`
        &:hover {
            color: ${props => props.theme.colourForeground};

            ${StyledLoader} {
                circle {
                    stroke: ${props => props.theme.colourForeground};
                }
            }
        }
    `}

    ${StyledLoader} {
        circle {
            stroke: ${props => props.theme.colourPrimary};
        }
    }
`;

const SecondaryButtonSmall = styled(SecondaryButton)`
    font-size: 1.4rem;
    height: 3rem;
    border-color: ${props => props.theme.colourDivider};

    ${Media.desktop`
        font-size: 1.4rem;
    `};
`;

const IconButton = styled(BaseButton)`
    background: ${props => props.theme.colourForegroundHover};
    border-color: ${props => props.isToggled ? props.theme.colourPrimary : props.theme.colourDivider};
    padding: 1.4rem 1.3rem;

    ${Media.desktop`
        padding: 1.4rem 1.3rem;

        &:hover {
            background: ${props => props.theme.colourForeground};

            svg {
                fill: ${props => props.theme.colourPrimary};
            }
        }
    `};

    /* Icons */
    svg {
        fill: ${props => props.theme.colourPrimary};
        height: 2rem;
        position: relative;
        width: 2rem;
    }
`;

const SmallIconButton = styled(IconButton)`
    height: 3rem;
    width: 3rem;

    ${Media.desktop`
        padding: 0.4rem;
        height: 3rem;
        width: 3rem;
    `};
`;

export default BaseButton;

export {
    BaseButton as Button,
    SecondaryButton,
    SecondaryButtonSmall,
    IconButton,
    SmallIconButton
};
