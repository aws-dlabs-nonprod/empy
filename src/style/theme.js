export default {
    textPrimary: '#2D373E',
    textSecondary: '#E7E6E6',
    textBody: '#1A1A1A',
    textDark: '#621A4B',

    colourPrimary: '#D5002B',
    colourPrimaryHover: '#e24d6b',
    colourPrimaryTransparent: 'rgba(0, 174, 236, 0.5)',

    colourForeground: '#FFFFFF',
    colourForegroundHover: '#F9F9F8',
    colourForegroundTransparent: 'rgba(255, 255, 255, 0.2)',

    colourInactive: '#7F7F7F',
    colourBackground: '#E5E5E5',
    colourDivider: '#D7D2CB',
    colourFaintBorder: '#8C9296',

    radius: '0.3rem',
    splitHeight: '50vh'
};
