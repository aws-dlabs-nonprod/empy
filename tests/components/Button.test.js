import React from 'react';
import { shallow, mount } from 'enzyme';
import 'jest-styled-components';
import theme from '@style/theme';

import Button from '@components/Button';
import Tip from '@components/ButtonTip';

describe('Button', () => {

    test('renders correctly', () => {
        const component = shallow(<Button theme={ theme } />);

        expect(component).toMatchSnapshot();
    });

    test('toggles correctly', () => {
        const component = mount(
            <Button
                theme={ theme }
                isToggled={ false }
                tip="off message"
                toggleTip="on message"
            />
        );

        expect(component.find(Tip).text()).toEqual('off message');

        component.setProps({ isToggled: true });

        expect(component.find(Tip).text()).toEqual('on message');
    });

    test('handles click', () => {
        const mockCallback = jest.fn();

        const component = shallow(
            <Button theme={ theme } onClick={ mockCallback } />
        );

        component.simulate('click');

        expect(mockCallback).toHaveBeenCalled();
    });

});
