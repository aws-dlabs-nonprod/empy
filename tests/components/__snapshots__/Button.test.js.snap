// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly 1`] = `
<StyledComponent
  forwardedComponent={
    Object {
      "$$typeof": Symbol(react.forward_ref),
      "attrs": Array [],
      "componentStyle": ComponentStyle {
        "componentId": "Button__BaseButton-sc-1t6ofhe-0",
        "isStatic": false,
        "rules": Array [
          "background:",
          [Function],
          ";border-radius:",
          [Function],
          ";box-sizing:border-box;color:",
          [Function],
          ";cursor:pointer;display:flex;justify-content:center;align-items:center;font-size:1.6rem;height:4.8rem;border:",
          [Function],
          ";margin:0 0 2rem 0;outline:none;padding:0 1.8rem;position:relative;width:100%;z-index:3;transition:background 0.2s ease,color 0.2s ease;",
          "@media (min-width:",
          "992",
          "px) and (min-height:",
          "0",
          "px){",
          "
        font-size: 1.8rem;
        padding: 0 2rem;
        width: 100%;

        &:hover {
            background: ",
          [Function],
          ";
            
            svg {
                fill: ",
          [Function],
          ";
            }

            ",
          ".ButtonTip__StyledButtonTip-zuccjc-0",
          " {
                opacity: ",
          [Function],
          ";
                transform: translateY(0);
            }
        }
    ",
          "}",
        ],
      },
      "displayName": "Button__BaseButton",
      "foldedComponentIds": Array [],
      "render": [Function],
      "styledComponentId": "Button__BaseButton-sc-1t6ofhe-0",
      "target": [Function],
      "toString": [Function],
      "usesTheme": true,
      "warnTooManyClasses": [Function],
      "withComponent": [Function],
    }
  }
  forwardedRef={null}
  theme={
    Object {
      "colourBackground": "#E5E5E5",
      "colourDivider": "#D7D2CB",
      "colourFaintBorder": "#8C9296",
      "colourForeground": "#FFFFFF",
      "colourForegroundHover": "#F9F9F8",
      "colourForegroundTransparent": "rgba(255, 255, 255, 0.2)",
      "colourInactive": "#7F7F7F",
      "colourPrimary": "#D5002B",
      "colourPrimaryHover": "#e24d6b",
      "colourPrimaryTransparent": "rgba(0, 174, 236, 0.5)",
      "radius": "0.3rem",
      "splitHeight": "50vh",
      "textBody": "#1A1A1A",
      "textDark": "#621A4B",
      "textPrimary": "#2D373E",
      "textSecondary": "#E7E6E6",
    }
  }
/>
`;
